### ISSUE TITLE
adding   `"sourceType": "module"` flags dynamic imports.

### ISSUE THREAD

**Issue by dkebler** [eslint/eslint#9811]
**Tell us about your environment**

* **ESLint Version:**  4.14
* **Node Version:**   9.3
* **npm Version:** 5.6

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```js
module.exports = {
  "env": {
    "es6": true,
    "node": true,
    "mocha": true
  },
  "parserOptions": {
    "ecmaVersion": 2017,
    "sourceType": "module"
  },
  "extends": "eslint:recommended",
  "rules": {
    "indent": [
      "error",
      2
    ],
    "no-console": 0,
    "semi": ["error", "never"],
    "linebreak-style": [
      "error",
      "unix"
    ],
    "quotes": [
      "error",
      "single"
    ]
  }
}
```
**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

added `    "sourceType": "module"`  to stop flagging `import`.   
That worked fine but....

**What did you expect to happen?**
not flag dynamic import line `await import`

**What actually happened? Please include the actual, raw output from ESLint.**
![image](https://user-images.githubusercontent.com/1316185/34642402-4546c27c-f2c7-11e7-82da-f997551e10f8.png)

**Parsing error: Unexpected token import (Fatal)**

Note: if I remove the sourceType: module then this stops flagging but of course then the standard import lines do get flagged.

run with atom linter-eslint 8.4 plugin  and  `@std/esm` by @jdalton which enables esm support including dynamic imports.

I understand this is all "new stuff" but come node 10 esm will be native (without using `@std/esm`) and 
I assume will include dynamic imports.

```js
import path from 'path';
import fs from 'fs';
// import { promisfy as pify } from 'util';
import deepset from 'lodash.set';
import walk from 'walk-filtered';

// const pStat = pify(fs.lstat);

const EXTENSIONS = ['.mjs'];

// rename to file name or remove default in a module
function reDefault(obj, fpath) {
  const newObj = {};
  Object.keys(obj).forEach((key) => {
    if (key !== 'default') newObj[key] = obj[key];
    else if (arguments.length === 2) newObj[path.parse(fpath).name] = obj.default;
  });
  return newObj;
}

async function importFile(directory, relativePath, options, results) {
  let module = await import(path.join(directory, relativePath));

  if (options.default === 'none') module = reDefault(module)
  else module = reDefault(module, relativePath)

  // merge result
  if (options.collect === 'mergeAll') Object.assign(results, module);
  else {
    const dirs = path.parse(relativePath).dir;
    if (dirs.length) deepset(results, dirs.replace(/\//g, '.'), module);
    else Object.assign(results, module);
  }
} // end importFile
```

**Comment by ljharb** [comment]
`import()` is stage 3; I don't think eslint supports it yet (sadly)

**Comment by jdalton** [comment]
@dkebler You'll want to use [babel-eslint](https://github.com/babel/babel-eslint).

**Comment by platinumazure** [comment]
As noted by @ljharb and @jdalton, ESLint (or rather, espree, the default ESLint parser) does not support dynamic import and you should use [babel-eslint](https://github.com/babel/babel-eslint).

ESLint does not support stage 3 features in general at this time.

Closing as the question appears to have been answered, but feel free to continue to discuss here or to raise any questions in a new issue or on our [Gitter channel](https://gitter.im/eslint/eslint).

**Comment by dkebler** [comment]
Just an observation about default parser import and import().   

Either way with souceType:module on or off eslint default parser catching `import` or `import()` keeps the linter from linting everything else :-(.   

So if I just don't want to go the babel-eslint route (which I don't) how can I get eslint default parser to at least ignore `import`, `export` and `import()` (i.e esm) so at least the rest of my code is linted?

**Comment by j-f1** [comment]
Unfortunately, that’s not how ESLint’s parser works. It’s either able to parse the file and pass it to ESLint, or it throws an error and fails to parse.

**Comment by ljharb** [comment]
In other words, `import()` is not a thing that exists in the language yet (at stage 4, according to acorn's unfortunate official policy), so it can't ignore it because it doesn't understand it in the first place.

**Comment by jdalton** [comment]
I mean it's ESLint's policy too. It's a real bummer it's shrugs all the way down.
